### PR TITLE
Feature/low level logs

### PR DESCRIPTION
**Pull request by a-zakir**

### CODE REVIEW

File: src/cpp/benders/benders_core/include/BendersMathLogger.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+#include <iostream>
+#include <list>
+#include <memory>
+
+#include "BendersStructsDatas.h"
+#include "common.h"
+const std::string MATHLOGGERCONTEXT = "Benders";
+
+inline std::string Indent(int size) { return std::string(size, ' '); }
+const std::string ITE = Indent(10) + "ITE";
```
Comment (tbittar) [inline src/cpp/benders/benders_core/include/BendersMathLogger.h]: As we have clean indents we may use `ITERATION` instead of `ITE` and adjust indentation

File: src/cpp/benders/benders_core/include/BendersMathLogger.h [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    Indent(15) + "SUB-PROBLEMS TIME (CPU)";
+const std::string SUB_PROBLEMS_TIME_WALL =
+    Indent(15) + "SUB-PROBLEMS TIME (WALL)";
+const std::string TIME_NOT_DOING_MASTER_OR_SUB_PROBLEMS_WALL =
+    Indent(15) + "TIME NOT DOING MASTER OR SUB-PROBLEMS (WALL)";
+class LogDestination {
+ public:
+  explicit LogDestination(std::ostream* stream) : stream_(stream) {}
+
+  // for std::endl
```
Comment (tbittar) [inline src/cpp/benders/benders_core/include/BendersMathLogger.h:27]: Remove commented code if useless

File: src/cpp/benders/benders_mpi/BendersMPI.cpp [hunk 3]
```diff
@@ -153,6 +155,8 @@ void BendersMpi::master_build_cuts(
   for (const auto &subproblem_data_map : gathered_subproblem_map) {
     for (auto &&[_, subproblem_data] : subproblem_data_map) {
       SetSubproblemCost(GetSubproblemCost() + subproblem_data.subproblem_cost);
+      // compute delta_cut >= options.CUT_MASTER_TOL;
```
Comment (tbittar) [inline src/cpp/benders/benders_mpi/BendersMPI.cpp:158]: This comment does not seem related to what happens here

File: src/cpp/benders/benders_mpi/BendersMPI.cpp [hunk 4]
```diff
@@ -274,6 +284,7 @@ void BendersMpi::PreRunInitialization() {
       OpenCsvFile();
     }
   }
+  mathLoggerDriver_->write_header();
```
Comment (tbittar) [inline src/cpp/benders/benders_mpi/BendersMPI.cpp:282]: Why do we need to call `write_header` twice ? Line 231 and 287

File: src/cpp/benders/factories/BendersFactory.cpp [hunk 5]
```diff
@@ -34,28 +35,35 @@ int RunBenders(char** argv, const std::filesystem::path& options_file,
     auto log_reports_name =
         std::filesystem::path(options.OUTPUTROOT) / "reportbenders.txt";
 
+    auto math_logs_file =
+        std::filesystem::path(options.OUTPUTROOT) / "math_log.txt";
```
Comment (tbittar) [inline src/cpp/benders/factories/BendersFactory.cpp]: Rename to `benders_solver_log`, this will be more meaningful for optimization people

File: src/cpp/benders/factories/include/LoggerFactories.h [hunk 6]
```diff
@@ -29,7 +31,32 @@ class FileAndStdoutLoggerFactory {
     masterLogger->addLogger(loggerUser);
     logger = masterLogger;
   }
-
   inline Logger get_logger() const { return logger; }
 };
+
+class MathLoggerFactory {
+ private:
+  MathLoggerDriver math_Logger_driver;
```
Comment (tbittar) [inline src/cpp/benders/factories/include/LoggerFactories.h]: Rename `math_logger_driver` instead of `math_Logger_driver` ?

File: src/cpp/benders/factories/include/LoggerFactories.h [hunk 7]
```diff
@@ -29,7 +31,32 @@ class FileAndStdoutLoggerFactory {
...
+      const BENDERSMETHOD &method, bool console_log,
+      const std::filesystem::path &math_logs_file_path = "") {
+    if (math_logs_file_path != "") {
+      auto math_logger_file =
+          std::make_shared<MathLoggerFile>(method, math_logs_file_path);
+      math_Logger_driver.add_logger(math_logger_file);
+    }
+
+      if (console_log) {
+        auto math_Logger_ostream = std::make_shared<MathLoggerOstream>(method);
```
Comment (tbittar) [inline src/cpp/benders/factories/include/LoggerFactories.h]: `math_logger_ostream` instead of `math_Logger_ostream` ?

File: src/cpp/benders/logger/MathLogger.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+void MathLogger::setHeadersList(const std::list<std::string>& headers) {
+  headers_.clear();
+  headers_ = headers;
+}
+
+void MathLogger::write_header() {
+  setHeadersList();
+  for (const auto& header : Headers()) {
+    TheLogDestination() << header;
```
Comment (tbittar) [inline src/cpp/benders/logger/MathLogger.cpp]: `LogDestination()` instead of `TheLogDestination()` seems more natural

File: src/cpp/benders/logger/MathLogger.cpp [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  }
+  TheLogDestination() << std::endl;
+}
+void MathLoggerBase::Print(const CurrentIterationData& data) {
+  TheLogDestination() << Indent(10) << data.it;
+  if (data.lb == -1e20)
+    TheLogDestination() << Indent(20) << "-INF";
```
Comment (tbittar) [inline src/cpp/benders/logger/MathLogger.cpp]: I prefer that we really print `1e+20` and `1e-20` in the math logs as these are the actual values of the data. the purpose of the math logs is to monitor exactly what happens in the data and not to have an interpretation of it as infinity

File: src/cpp/benders/logger/MathLogger.cpp [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+void MathLoggerBendersByBatch::setHeadersList() {
+  MathLogger::setHeadersList({ITE, LB, MINSIMPLEX, MAXSIMPLEX, TIMEMASTER,
+                              SUB_PROBLEMS_TIME_CPU, SUB_PROBLEMS_TIME_WALL,
+                              TIME_NOT_DOING_MASTER_OR_SUB_PROBLEMS_WALL});
+}
+
+void MathLoggerBendersByBatch::Print(const CurrentIterationData& data) {
+  TheLogDestination() << Indent(10) << data.it;
+  if (data.lb == -1e20)
```
Comment (tbittar) [inline src/cpp/benders/logger/MathLogger.cpp]: Same as above

File: src/cpp/benders/logger/MathLogger.cpp [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+  TheLogDestination() << Indent(15) << std::setprecision(2)
+                      << data.subproblems_walltime;
+
+  TheLogDestination() << std::endl;
+}
+
+void MathLoggerBendersByBatch::setHeadersList() {
+  MathLogger::setHeadersList({ITE, LB, MINSIMPLEX, MAXSIMPLEX, TIMEMASTER,
+                              SUB_PROBLEMS_TIME_CPU, SUB_PROBLEMS_TIME_WALL,
+                              TIME_NOT_DOING_MASTER_OR_SUB_PROBLEMS_WALL});
```
Comment (tbittar) [inline src/cpp/benders/logger/MathLogger.cpp]: Can we also add this log to the base case ?

File: src/cpp/benders/logger/MathLogger.cpp [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+  // TheLogDestination() << Indent(15) << data.deletedcut;
+  TheLogDestination() << Indent(15) << std::setprecision(2)
+                      << data.timer_master;
+  TheLogDestination() << Indent(15) << std::setprecision(2)
+                      << data.subproblems_cumulative_cputime;
+  TheLogDestination() << Indent(15) << std::setprecision(2)
+                      << data.subproblems_walltime;
+
+  TheLogDestination() << std::endl;
```
Comment (tbittar) [inline src/cpp/benders/logger/MathLogger.cpp]: Add missing data

File: src/cpp/benders/benders_mpi/BendersMPI.cpp (hunk 3 above)
Comment (a-zakir) [reply src/cpp/benders/benders_mpi/BendersMPI.cpp:158]: yes it's a reminder for a specific data....

File: src/cpp/benders/benders_mpi/BendersMPI.cpp (hunk 4 above)
Comment (a-zakir) [reply src/cpp/benders/benders_mpi/BendersMPI.cpp:282]: l 231 is the old version, i'll clean

File: src/cpp/benders/logger/MathLogger.cpp (hunk 9 above)
Comment (a-zakir) [reply src/cpp/benders/logger/MathLogger.cpp]: you say, i do

File: src/cpp/benders/logger/MathLogger.cpp (hunk 11 above)
Comment (a-zakir) [reply src/cpp/benders/logger/MathLogger.cpp]: coming...

File: src/cpp/benders/logger/MathLogger.cpp (hunk 8 above)
Comment (a-zakir) [reply src/cpp/benders/logger/MathLogger.cpp]: i've a struct named LogDestination, i suggest LogsDestination
